File: Advent2022/Day1/Day1.py
from typing import Generator

import aoc_utils as au


def read_elves(foods: list[str]) -> Generator[int, None, None]:

    calories = 0
    for food in foods:
        if food == "":
            yield calories
            calories = 0
        else:
            calories += int(food)


# expected 24000
def test_one(foods: list[str]) -> int:
    return max(read_elves(foods))


# expected 67016
def part_one(foods: list[str]) -> int:
    return max(read_elves(foods))


# expected 45000
def test_two(foods: list[str]) -> int:
    return sum(sorted(read_elves(foods))[-3:])


# expected 200116
def part_two(foods: list[str]) -> int:
    return sum(sorted(read_elves(foods))[-3:])


if __name__ == "__main__":
    day = au.Day(
        2022,
        1,
        test_one,
        test_two,
        part_one,
        part_two,
        input=au.Day.InType.INPUT_MULTI_LINE_STR,
        test_input=au.Day.InType.INPUT_MULTI_LINE_STR,
    )

    day.run_all(run_tests=True)

File: Advent2022/Day1/Day1.cpp
#include "aoc_utils.h"

#include <string>
#include <vector>
#include <algorithm>
#include <numeric>

using namespace std;

class Day1 : public au::Day<vector<string>>
{
private:
    vector<int> collapse(const vector<string> &input)
    {
        vector<int> rc;

        int calories = 0;
        for (auto line : input)
        {
            if (line.empty())
            {
                rc.push_back(calories);
                calories = 0;
            }
            else
                calories += stoi(line);
        }

        rc.push_back(calories);
        return rc;
    }

public:
    Day1() : Day(2022, 1) {}

    int test_one(const vector<string> &foods)
    {
        const vector<int> elves = collapse(foods);
        return *max_element(elves.begin(), elves.end());
    }

    int part_one(const vector<string> &foods)
    {
        const vector<int> elves = collapse(foods);
        return *max_element(elves.begin(), elves.end());
    }

    int test_two(const vector<string> &foods)
    {
        vector<int> elves = collapse(foods);
        sort(elves.begin(), elves.end());
        return accumulate(elves.end() - 3, elves.end(), 0);
    }

    int part_two(const vector<string> &foods)
    {
        vector<int> elves = collapse(foods);
        sort(elves.begin(), elves.end());
        return accumulate(elves.end() - 3, elves.end(), 0);
    }
};

int main()
{
    Day1().run_all(true);

    return 0;
}